modules/react/common/lib/utils/components.ts

```diff
@@ -78,26 +76,26 @@ type ExtractHTMLAttributes<
  * @example
  * interface MyProps extends ExtractProps<typeof Card.Body> {}
  *
- * ExtractProps<typeof Card>; // React.ClassAttributes<HTMLDivElement> & React.HTMLAttributes<HTMLDivElement> & CardProps
- * ExtractProps<typeof Card, 'aside'>; // React.ClassAttributes<HTMLElement> & React.HTMLAttributes<HTMLElement> & CardProps
+ * ExtractProps<typeof Card>; // React.HTMLAttributes<HTMLDivElement> & CardProps
+ * ExtractProps<typeof Card, 'aside'>; // React.HTMLAttributes<HTMLElement> & CardProps
  * ExtractProps<typeof Card, never>; // CardProps
  */
 export type ExtractProps<
   TComponent,
   TElement extends keyof JSX.IntrinsicElements | undefined | never = undefined
 > = TComponent extends ElementComponent<infer E, infer P> // test if `TComponent` is an `ElementComponent`, while inferring both default element and props associated
   ? E extends keyof JSX.IntrinsicElements // test if the inferred element `E` is in JSX.IntrinsicElements
-    ? [TElement] extends [never] // test if user passed `never` for the `TElement` override. https://github.com/microsoft/TypeScript/issues/23182
-      ? P // if `TElement` was `never`, return only the inferred props `P`
+    ? TElement extends keyof JSX.IntrinsicElements
+      ? ExtractHTMLAttributes<JSX.IntrinsicElements[TElement]> & P // `TElement` was defined, so we'll return the HTML attribute interface + inferred `P` props
       : TElement extends undefined // else test if TElement was defined
       ? ExtractHTMLAttributes<JSX.IntrinsicElements[E]> & P // `TElement` wasn't explicitly defined, so let's fall back to the inferred element's HTML attribute interface + props `P`
-      // TS thinks `TElement` cannot be an index type because of the `undefined`, but we've already ensured that can't happen, so we'll ignore that error
-      // @ts-ignore
-      : ExtractHTMLAttributes<JSX.IntrinsicElements[TElement]> & P // `TElement` was defined, so we'll return the HTML attribute interface + inferred `P` props
+      : P // `TElement` was `never`, return only the inferred props `P`
     : P // E isn't in JSX.IntrinsicElements, we don't know what it is, so return the inferred props `P`
   : TComponent extends Component<infer P> // test if `TComponent` is a `Component`, while inferring props `P`
   ? P // it was a `Component`, return inferred props `P`
-  : {}; // `TComponent` does not extend either `ElementComponent` or `Component`, return an empty object
+  : TComponent extends React.ComponentType<infer P> // test if `TComponent` is a `React.ComponentType` (class or functional component)
+  ? P // it was a `React.ComponentType`, return inferred props `P`
+  : {}; // We don't know what `TComponent` was, return an empty object
```

Comment:
I'm sneaking this feature in. `ExtractProps` can now extract the props of any React component. Why not?